### PR TITLE
Added M-x ensime-troubleshooting

### PR DESCRIPTION
**Pull request by d1egoaz**
Removed duplicated optional in https://github.com/ensime/ensime-emacs/pull/457

Fixes https://github.com/ensime/ensime-emacs/issues/364

- I'd like to know what's the best place to put the function? (currently it's in ensime-mode.el)
- Also is not adding the report content by default to the kill-ring, only if universal argument is provided.
Should be the contrary? added by default to kill-ring and if universal argument is provided just print the content to *messages* buffer?
- Can you try if keymap combination is working?

### READ THIS
Have you read all the documentation at [Ensime Emacs](http://ensime.org/editors/emacs/)? Please do, we put a lot of effort into it.
Most problems can be resolved easily by following a simple process, follow through our
[Troubleshooting guide](http://ensime.org/editors/emacs/troubleshooting/) and share the debugging information below as a gist if you need to escalate.
Do not paste this directly into the gitter chat room as it is very verbose.
*****************

## System Information - Ensime troubleshoot report

- OS: `darwin`
- Emacs: `24.5.1`
- ensime-sbt-command: `nil`
- debug-on-error: `nil`
- exec-path: `(/Users/d1egoaz/.rbenv/shims/ ~/bin/ /Users/d1egoaz/.rbenv/bin/ /usr/local/bin/ /usr/local/bin/ /usr/bin/ /bin/ /usr/sbin/ /sbin/ /opt/X11/bin/ /Users/d1egoaz/.rbenv/shims/ ~/bin/ /Users/d1egoaz/.rbenv/bin/ /usr/local/git/bin/ /usr/local/sbin/ /Library/Java/JavaVirtualMachines/jdk1.8.0_45.jdk/Contents/Home/bin/ /Users/d1egoaz/.rbenv/shims/ /usr/bin/ /bin/ /usr/sbin/ /sbin/ /usr/local/bin/ /usr/local/git/bin/ /usr/local/sbin/ /Library/Java/JavaVirtualMachines/jdk1.8.0_45.jdk/Contents/Home/bin/ /usr/local/Cellar/emacs-plus/24.5/libexec/emacs/24.5/x86_64-apple-darwin14.4.0/)`
- Backtrace:
```
Debugger entered--Lisp error: (quit)
  (while t (setq n (+ n 1)))
  (let ((n 0)) (while t (setq n (+ n 1))))
  eval-region(18252 18299)  ; Reading at buffer position 18299
  call-interactively(eval-region nil nil)
  command-execute(eval-region)

```
- .ensime file content:
```
(
 :root-dir "/Users/d1egoaz/dev/scala/akkainfo/hakky-hour"
 :cache-dir "/Users/d1egoaz/dev/scala/akkainfo/hakky-hour/.ensime_cache"
 :scala-compiler-jars ("/Users/d1egoaz/.ivy2/cache/org.scala-lang/scala-compiler/jars/scala-compiler-2.11.2.jar" "/Users/d1egoaz/.ivy2/cache/org.scala-lang/scala-library/jars/scala-library-2.11.2.jar" "/Users/d1egoaz/.ivy2/cache/org.scala-lang/scala-reflect/jars/scala-reflect-2.11.2.jar" "/Users/d1egoaz/.ivy2/cache/org.scala-lang/scalap/jars/scalap-2.11.2.jar")
 :name "hakkyHour"
 :java-home "/Library/Java/JavaVirtualMachines/jdk1.8.0_45.jdk/Contents/Home"
 :java-flags ("-Xss2m" "-Xms1024m" "-Xmx1024m" "-XX:ReservedCodeCacheSize=128m" "-XX:MaxMetaspaceSize=256m" "-Dfile.encoding=UTF8")
 :java-compiler-args nil
 :reference-source-roots ("/Library/Java/JavaVirtualMachines/jdk1.8.0_45.jdk/Contents/Home/src.zip")
 :scala-version "2.11.2"
 :compiler-args ("-unchecked" "-deprecation" "-language:_" "-target:jvm-1.6" "-encoding" "UTF-8" "-feature" "-Xlint" "-Yinline-warnings" "-Yno-adapted-args" "-Ywarn-dead-code" "-Ywarn-numeric-widen" "-Xfuture" "-Ywarn-unused-import" "-Ymacro-expand:discard")
 :formatting-prefs nil
 :disable-source-monitoring nil
 :disable-class-monitoring nil
 :subprojects ((
   :name "hakkyHour"
   :source-roots ("/Users/d1egoaz/dev/scala/akkainfo/hakky-hour/src/test/extra" "/Users/d1egoaz/dev/scala/akkainfo/hakky-hour/src/main/scala" "/Users/d1egoaz/dev/scala/akkainfo/hakky-hour/src/test/scala")
   :targets ("/Users/d1egoaz/dev/scala/akkainfo/hakky-hour/target/scala-2.11/classes")
   :test-targets ("/Users/d1egoaz/dev/scala/akkainfo/hakky-hour/target/scala-2.11/test-classes")
   :depends-on-modules nil
   :compile-deps ("/Users/d1egoaz/.ivy2/cache/com.typesafe.akka/akka-actor_2.11/jars/akka-actor_2.11-2.3.6.jar" "/Users/d1egoaz/.ivy2/cache/com.typesafe.akka/akka-slf4j_2.11/jars/akka-slf4j_2.11-2.3.6.jar" "/Users/d1egoaz/.ivy2/cache/com.typesafe/config/bundles/config-1.2.1.jar" "/Users/d1egoaz/.ivy2/cache/ch.qos.logback/logback-classic/jars/logback-classic-1.1.2.jar" "/Users/d1egoaz/.ivy2/cache/ch.qos.logback/logback-core/jars/logback-core-1.1.2.jar" "/Users/d1egoaz/.ivy2/cache/org.scala-lang/scala-library/jars/scala-library-2.11.2.jar" "/Users/d1egoaz/.ivy2/cache/org.scala-lang.modules/scala-parser-combinators_2.11/bundles/scala-parser-combinators_2.11-1.0.2.jar" "/Users/d1egoaz/.ivy2/cache/org.slf4j/slf4j-api/jars/slf4j-api-1.7.6.jar")
   :runtime-deps nil
   :test-deps ("/Users/d1egoaz/.ivy2/cache/com.typesafe.akka/akka-testkit_2.11/jars/akka-testkit_2.11-2.3.6.jar" "/Users/d1egoaz/.ivy2/cache/org.scala-lang/scala-reflect/jars/scala-reflect-2.11.2.jar" "/Users/d1egoaz/.ivy2/cache/org.scala-lang.modules/scala-xml_2.11/bundles/scala-xml_2.11-1.0.2.jar" "/Users/d1egoaz/.ivy2/cache/org.scalatest/scalatest_2.11/bundles/scalatest_2.11-2.2.2.jar")
   :doc-jars ("/Users/d1egoaz/.ivy2/cache/com.typesafe.akka/akka-actor_2.11/docs/akka-actor_2.11-2.3.6-javadoc.jar" "/Users/d1egoaz/.ivy2/cache/com.typesafe.akka/akka-slf4j_2.11/docs/akka-slf4j_2.11-2.3.6-javadoc.jar" "/Users/d1egoaz/.ivy2/cache/com.typesafe.akka/akka-testkit_2.11/docs/akka-testkit_2.11-2.3.6-javadoc.jar" "/Users/d1egoaz/.ivy2/cache/com.typesafe/config/docs/config-1.2.1-javadoc.jar" "/Users/d1egoaz/dev/scala/akkainfo/hakky-hour/target/scala-2.11/hakky-hour_2.11-4.0.0-javadoc.jar" "/Users/d1egoaz/.ivy2/cache/ch.qos.logback/logback-classic/docs/logback-classic-1.1.2-javadoc.jar" "/Users/d1egoaz/.ivy2/cache/ch.qos.logback/logback-core/docs/logback-core-1.1.2-javadoc.jar" "/Users/d1egoaz/.ivy2/cache/org.scala-lang/scala-library/docs/scala-library-2.11.2-javadoc.jar" "/Users/d1egoaz/.ivy2/cache/org.scala-lang.modules/scala-parser-combinators_2.11/docs/scala-parser-combinators_2.11-1.0.2-javadoc.jar" "/Users/d1egoaz/.ivy2/cache/org.scala-lang/scala-reflect/docs/scala-reflect-2.11.2-javadoc.jar" "/Users/d1egoaz/.ivy2/cache/org.scala-lang.modules/scala-xml_2.11/docs/scala-xml_2.11-1.0.2-javadoc.jar" "/Users/d1egoaz/.ivy2/cache/org.scalatest/scalatest_2.11/docs/scalatest_2.11-2.2.2-javadoc.jar" "/Users/d1egoaz/.ivy2/cache/org.slf4j/slf4j-api/docs/slf4j-api-1.7.6-javadoc.jar")
   :reference-source-roots ("/Users/d1egoaz/.ivy2/cache/com.typesafe.akka/akka-actor_2.11/srcs/akka-actor_2.11-2.3.6-sources.jar" "/Users/d1egoaz/.ivy2/cache/com.typesafe.akka/akka-slf4j_2.11/srcs/akka-slf4j_2.11-2.3.6-sources.jar" "/Users/d1egoaz/.ivy2/cache/com.typesafe.akka/akka-testkit_2.11/srcs/akka-testkit_2.11-2.3.6-sources.jar" "/Users/d1egoaz/.ivy2/cache/com.typesafe/config/srcs/config-1.2.1-sources.jar" "/Users/d1egoaz/.ivy2/cache/ch.qos.logback/logback-classic/srcs/logback-classic-1.1.2-sources.jar" "/Users/d1egoaz/.ivy2/cache/ch.qos.logback/logback-core/srcs/logback-core-1.1.2-sources.jar" "/Users/d1egoaz/.ivy2/cache/org.scala-lang/scala-library/srcs/scala-library-2.11.2-sources.jar" "/Users/d1egoaz/.ivy2/cache/org.scala-lang.modules/scala-parser-combinators_2.11/srcs/scala-parser-combinators_2.11-1.0.2-sources.jar" "/Users/d1egoaz/.ivy2/cache/org.scala-lang/scala-reflect/srcs/scala-reflect-2.11.2-sources.jar" "/Users/d1egoaz/.ivy2/cache/org.scala-lang.modules/scala-xml_2.11/srcs/scala-xml_2.11-1.0.2-sources.jar" "/Users/d1egoaz/.ivy2/cache/org.scalatest/scalatest_2.11/srcs/scalatest_2.11-2.2.2-sources.jar" "/Users/d1egoaz/.ivy2/cache/org.slf4j/slf4j-api/srcs/slf4j-api-1.7.6-sources.jar")))
)

```

### CODE REVIEW

**Comment by fommil** [comment]
paste into a **gist** not the gitter room

File: ensime-mode.el [hunk 1]
```diff
@@ -214,6 +216,7 @@
     ["Go to stacktrace buffer" ensime-stacktrace-switch]
     ["Go to Scala REPL" ensime-inf-switch]
     ["Shutdown ENSIME server" ensime-shutdown]
+    ["Ensime Troubleshooting" ensime-troubleshooting]
```
Comment (fommil) [inline ensime-mode.el]: no need to say "Ensime" here

**Comment by fommil** [comment]
this is pretty cool! Wording could do with a little bit of changing, but the concept is great.

File: ensime-mode.el (hunk 1 above)
Comment (d1egoaz) [inline ensime-mode.el]: ok

**Comment by fommil** [comment]
"Have you read all the documentation at ensime.org/editors/emacs? Please do, we put a lot of effort into it.

Most problems can be resolved easily by following a simple process, follow through our http://ensime.org/editors/emacs/troubleshooting/ guide and share the debugging information below as a gist if you need to escalate. Do not paste this directly into the gitter chat room as it is very verbose."

**Comment by fommil** [comment]
ensime-mode is fine, I like the idea of adding to the kill ring.

**Comment by fommil** [comment]
@d1egoaz another thing like this that would be useful would be a welcome screen for new users pointing out where to find the documentation and letting them know about `M-x ensime-troubleshooting` at http://ensime.org and would have to be disabled with a setting `(setq ensime-show-introduction nil)`.

**Comment by fommil** [comment]
I wouldn't bother adding a key combination for this btw, `M-x ensime-troubleshooting` and a menu entry is all that is needed.

**Comment by d1egoaz** [comment]
Fixed initial comment and added by default to kill ring.

File: ensime-mode.el [hunk 2]
```diff
@@ -472,13 +472,14 @@
 
 (defun ensime-troubleshooting (arg)
   "Information about emacs and ensime installation.
-A prefix argument will add the report to the kill ring."
+A prefix argument will prevent to add the report to the kill ring."
```
Comment (fommil) [inline ensime-mode.el]: this is overengineering, no need to handle an alt case, let's just add to the kill ring.

File: ensime-mode.el [hunk 3]
```diff
@@ -472,13 +472,14 @@
...
   (interactive "P")
   (let ((info (format
                (concat "\n*** READ THIS ***\n"
-                       "Before posting this to Gitter (https://gitter.im/ensime/ensime-emacs)\n"
-                       "Have you read and follow the steps detailed in the troubleshooting guide http://ensime.github.io/editors/emacs/troubleshooting/ ?\n"
-                       "If yes, please paste this info on the Gitter channel, otherwise please read the guide first\n"
+                       "Have you read all the documentation at ensime.org/editors/emacs? Please do, we put a lot of effort into it.\n"
```
Comment (fommil) [inline ensime-mode.el]: I think you caught an old version of this before I edited it for typos

**Comment by d1egoaz** [comment]
@fommil can we do the welcome screen as another PR? I'll look some free time this weekend

File: ensime-mode.el [hunk 4]
```diff
@@ -505,9 +506,10 @@ A prefix argument will add the report to the kill ring."
                          (insert-file-contents filePath)
                          (buffer-string))))
                  (progn "Not provided")))))
-    (when arg
+    (unless arg
       (kill-new info))
-    (message info)))
+    (message info)
```
Comment (fommil) [inline ensime-mode.el]: maybe we want to put this in a buffer of its own? Like `*ensime-troubleshooting*`. In fact, if we do that there is no reason to add it to the kill ring.

File: ensime-mode.el (hunk 3 above)
Comment (d1egoaz) [inline ensime-mode.el]: Will fix!

File: ensime-mode.el (hunk 2 above)
Comment (d1egoaz) [inline ensime-mode.el]: k

File: ensime-mode.el (hunk 4 above)
Comment (d1egoaz) [inline ensime-mode.el]: Sure, but it has to be tomorrow, I need to read first about creating buffers in emacs 😆

Comment (fommil) [inline ensime-mode.el]: `with-buffer` I think

**Comment by d1egoaz** [comment]
@fommil updated and squashed, added output to a temporary buffer

**Comment by fommil** [comment]
Cool! Btw no need to squash anymore since github does this now

**Comment by fommil** [comment]
thanks! please update the docs. What next? :smile:

**Pull request merged**